### PR TITLE
[Swift bridging] Consistently use correctly-defined Swift(U)Int for bridging

### PR DESCRIPTION
**Pull request by DougGregor**
The `SwiftInt` and `SwiftUInt` types in the compiler provide `C` equivalents to Swift's `Int` and `UInt`. Consistently use these types in the C/Swift bridging layer, to avoid painful truncations and data-size mismatches.

This is harder than it should be. The canonical C types are `ptrdiff_t` and `size_t`, but we're in the depths of the compiler we can't include the C library headers that provide them because they introduce cyclic module dependencies. Sigh.

SwiftShims has some logic to compute these types. However, SwiftShims is part of the Swift runtime, not the compiler, so those headers cannot be included here. So, we clone the logic and simplify it somewhat for our use case.

This fixes truncation issues on Windows, where the uses of `unsigned long` and `long` for Swift(U)Int are incorrect.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CBasicBridging.h [hunk 1]
```diff
@@ -21,6 +21,64 @@
...
+#define __swiftc_intn_t(n) __swiftc_join3(__swiftc_int, n, _t)
+#define __swiftc_uintn_t(n) __swiftc_join3(__swiftc_uint, n, _t)
+
+#if defined(_MSC_VER) && !defined(__clang__)
+#if defined(_WIN32)
+typedef __swiftc_int32_t SwiftInt;
+typedef __swiftc_uint32_t SwiftUInt;
+#elif defined(_WIN64)
+typedef __swiftc_int64_t SwiftInt;
+typedef __swiftc_uint64_t SwiftUInt;
```
Comment (compnerd) [inline include/swift/Basic/CBasicBridging.h:72]: This is inverted ... `_WIN64` will have `_WIN32` defined IIRC, so we will always create 32-bit typedefs.

Comment (DougGregor) [reply include/swift/Basic/CBasicBridging.h:72]: Oh no! This comes out of SwiftShims. I can fix in both places, but I don't have a good way to test the fix

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CBasicBridging.h (hunk 1 above)
Comment (DougGregor) [reply include/swift/Basic/CBasicBridging.h:69]: Fixed in https://github.com/apple/swift/pull/68278/commits/592752741917c0ff76e382646f6d3e8a0bdd30c1, thank you so much!

**Review by compnerd — APPROVED** [review]
Awesome!  Thanks again @DougGregor!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CBasicBridging.h [hunk 2]
```diff
@@ -21,6 +21,69 @@
...
+typedef __swiftc_int32_t SwiftInt;
+typedef __swiftc_uint32_t SwiftUInt;
+#else
+#error unknown windows pointer width
+#endif
+#else
+typedef __swiftc_intn_t(__INTPTR_WIDTH__) SwiftInt;
+typedef __swiftc_uintn_t(__INTPTR_WIDTH__) SwiftUInt;
+#endif
+#endif
```
Comment (compnerd) [inline include/swift/Basic/CBasicBridging.h:85]:
```suggestion
```

File: include/swift/Basic/CBasicBridging.h [hunk 3]
```diff
@@ -21,6 +21,69 @@
...
+typedef uint32_t __swiftc_uint32_t;
+typedef intptr_t __swiftc_intptr_t;
+typedef uintptr_t __swiftc_uintptr_t;
+#else
+typedef __INT64_TYPE__ __swiftc_int64_t;
+#ifdef __UINT64_TYPE__
+typedef __UINT64_TYPE__ __swiftc_uint64_t;
+#else
+typedef unsigned __INT64_TYPE__ __swiftc_uint64_t;
+#endif
```
Comment (compnerd) [inline include/swift/Basic/CBasicBridging.h:57]:
```suggestion
#endif
#endif
```

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CBasicBridging.h [hunk 4]
```diff
@@ -21,6 +21,69 @@
...
+#endif
```
Comment (compnerd) [inline include/swift/Basic/CBasicBridging.h:58]:
```suggestion
```

File: include/swift/Basic/CBasicBridging.h [hunk 5]
```diff
@@ -21,6 +21,69 @@
...
+
+typedef __INT32_TYPE__ __swiftc_int32_t;
+#ifdef __UINT32_TYPE__
+typedef __UINT32_TYPE__ __swiftc_uint32_t;
+#else
+typedef unsigned __INT32_TYPE__ __swiftc_uint32_t;
+#endif
```
Comment (compnerd) [inline include/swift/Basic/CBasicBridging.h:65]:
```suggestion
#endif
#endif
```

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Pull request merged**